### DISCUSSION TITLE
Creating, reading, removing files

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [tauri-apps/tauri#6380] [Q&A] [answered]
Hello,

It's me again, new day, new feature! 😁 

I am trying to implement file management in my app. I tried following https://www.matthewtao.com/blog/post/glipma-devlog-2/ and got to:

On the JS side
```
const DocumentContainer = ({ position }: IProps) => {
 ...
  useEffect(() => {
    const createFolder = async () => {
      await createDir(getFolderName(position), {
        dir: BaseDirectory.AppData,
        recursive: true,
      });
    };
    exists(getFolderName(position), { dir: BaseDirectory.AppData }).then(
      (result) => {
        if (!result) {
          createFolder();
        }
      }
    );
  }, [position]);
  const onDrop = (acceptedFiles: File[]) => {
    console.log(acceptedFiles);
    acceptedFiles.forEach((file) => {
      try {
        const path = `${getFolderName(position)}/${file.name}`;
        const reader = new FileReader();
        reader.onabort = () => console.log('file reading was aborted')
        reader.onerror = () => console.log('file reading has failed')
        reader.onload = async () => {
          // Do whatever you want with the file contents
          const binaryStr = reader.result
          await writeBinaryFile({
            path: path,
            contents: binaryStr as BinaryFileContents
          }, {
            dir: BaseDirectory.AppData
          });
          await create({
            name: file.name,
            path,
            size: file.size,
            positionId: position.id
          });
        }
        reader.readAsArrayBuffer(file);
      } catch (e) {
        console.error(e);
      }
    });
  };
  ...
  return ...;
};
```
I added to my `tauri.conf.json`:
```
...
    "tauri": {
    "allowlist": {
      "fs": {
        "all": true,
        "scope": ["$APPDATA/*"]
      },
      "path": {
        "all": true
      }
    },
...
```

From what I read, this should give access to my app to the app_data directory, with `$APPDATA` being replaced at runtime.

But when I run this code, I get errors stating:
`Unhandled Promise Rejection: path not allowed on the configured scope: /home/tristan/.local/share/com.tristanpouliquen.dev/positions/4` when trying to create the directory
`Unhandled Promise Rejection: path not allowed on the configured scope: /home/tristan/.local/share/com.tristanpouliquen.dev/positions/4/8R44557972694.pdf` when trying to create a file

From what I understand, my Tauri app seems to not have the scope needed to interact with my folders.

What am I missing in my `tauri.conf.json`?

**Comment by FabianLars** [comment, marked as answer]
Try `"$APPDATA/**"` (double `*`) there was a fix related to a security issue and now `*` and `**` behave differently iirc (meaning the blog post is a bit out of date about that)

**Reply by TristanPouliquen** [reply]
As always my savior! 😁 

Maybe this doc should have a snippet about this too then: https://tauri.app/v1/api/js/fs/

**Reply by FabianLars** [reply]
True! Even the one example at the top may be misleading since it uses `*` too (it's not always wrong but we at least should explain the difference) without showing the other.

**Comment by itrumper** [comment]
I have tried the double `*` and am getting a slightly different error:
`Unhandled Promise Rejection: path: /Users/isaac/Library/Application Support/com.eleoptics.ember/license-key.txt: No such file or directory (os error 2)`

Which is true, this path doesn't exist. Is that expected during development? I'm running macOS 13.2.1.

As a workaround, I tried to create the directory using
`await createDir('', { dir: BaseDirectory.AppData, recursive: true });`
Which yielded the same error as the original post, that the path is not allowed on the configured scope.

My `tauri.conf.json`:
```
"tauri": {
    "allowlist": {
      "fs": {
        "all": true,
        "scope": [
          "$RESOURCE/**",
          "$APPDATA/**"
        ]
      },
  }
}
```

**Reply by FabianLars** [reply]
> Which is true, this path doesn't exist. Is that expected during development?

It's expected in development _and_ production. Tauri doesn't create any dirs automatically unless it needs it itself too.

> Which yielded the same error as the original post, that the path is not allowed on the configured scope.

Try adding the folder itself to the scope: `["$APPDATA/", "$APPDATA/**"]` - `**` only allows the children of the folder, not the folder itself.

**Reply by itrumper** [reply]
> It's expected in development and production. Tauri doesn't create any dirs automatically unless it needs it itself too.

Ah, I was confused about this behavior when it came to the named directories like `$APPDATA`. I thought those were created on install. Thank you for the clarification.

> Try adding the folder itself to the scope: ["$APPDATA/", "$APPDATA/**"] - ** only allows the children of the folder, not the folder itself.

With this additional info, I think I finally understand how the Tauri filesystem works! Thanks :)
